### PR TITLE
feat: ComboBox component

### LINKED ISSUE #170: New component: Combo box
# [Combo Box](https://designsystem.digital.gov/components/form-controls/#combo-box)
A combo box helps users select an item from a large list of options.

## Proposed Requirements
```
Props = {
  id: string,
  name: string
  children: React.ReactNode
  options: [],
  onSelect: (option) => void,
  defaultValue? : string
  required?: boolean
} & JSX.IntrinsicElements['select']
``` 
- base element is `<select>`
- ability clear selections or already typed text
- See [uswds combo box js](https://github.com/uswds/uswds/blob/develop/src/js/components/combo-box.js)

### PR DESCRIPTION
**Pull request by haworku**
# Summary
USWDS component: https://designsystem.digital.gov/form-controls/03-combo-box/

Current Status:
- ✅ basic functionality
     - Main component code is `ComboBox.tsx` and `ComboBox.test.tsx`. 
     - Includes storybook examples (`ComboBox.stories.tsx`) and implementation in example app (`/example` files). 
-  🔎 5 tests are skipped with comments. 
    - All represent valid minor bugs🐛, most relating to tab and hover behavior in the options dropdown.
    - _Should we consider these blockers to merging this first pass?_

## Related Issues or PRs
closes #170 

## How To Test
- Forms > ComboBox in storybook

### Screenshots (optional)
Screen Shot 2020-12-07 at 12 19 31 PM

### CODE REVIEW

File: src/components/forms/ComboBox/ComboBox.test.tsx [hunk 1]
```diff
@@ -0,0 +1,1111 @@
...
+      const { getByTestId } = render(
+        <ComboBox
+          id="favorite-fruit"
+          name="favorite-fruit"
+          options={fruitOptions}
+          onChange={jest.fn()}
+        />
+      )
+
+      userEvent.type(getByTestId('combo-box-toggle'), 'app')
```
Comment (suzubara) [inline src/components/forms/ComboBox/ComboBox.test.tsx:382]: is this meant to be `combo-box-input`?

File: src/components/forms/ComboBox/ComboBox.test.tsx [hunk 2]
```diff
@@ -0,0 +1,1111 @@
...
+          onChange={jest.fn()}
+        />
+      )
+
+      userEvent.type(getByTestId('combo-box-input'), 'b')
+      expect(getByTestId('combo-box-input')).toHaveFocus()
+
+      userEvent.click(getByTestId('combo-box-clear-button'))
+      fireEvent.blur(getByTestId('combo-box-clear-button'))
+      expect(getByTestId('combo-box-input')).toHaveFocus()
```
Comment (suzubara) [inline src/components/forms/ComboBox/ComboBox.test.tsx:483]: should the clear button be tabbable? I wasn't able to in chrome

File: src/components/forms/ComboBox/ComboBox.test.tsx [hunk 3]
```diff
@@ -0,0 +1,1111 @@
...
+        />
+      )
+
+      fireEvent.click(getByTestId('combo-box-input'))
+      fireEvent.focus(getByTestId('combo-box-option-yuzu'))
+      fireEvent.keyDown(getByTestId('combo-box-option-yuzu'), {
+        key: 'ArrowDown',
+      })
+
+      expect(getByTestId('combo-box-option-yuzu')).toHaveFocus()
```
Comment (suzubara) [inline src/components/forms/ComboBox/ComboBox.test.tsx:723]: fwiw, in Chrome when I press the down arrow after the last option it does not change focus

File: src/components/forms/ComboBox/ComboBox.test.tsx [hunk 4]
```diff
@@ -0,0 +1,1111 @@
...
+      )
+
+      // Apple is the item at top of list
+      fireEvent.focus(getByTestId('combo-box-option-apple'))
+      fireEvent.keyDown(getByTestId('combo-box-option-apple'), {
+        key: 'ArrowUp',
+      })
+
+      expect(getByTestId('combo-box-option-apple')).toHaveFocus()
+      expect(getByTestId('combo-box-option-list')).toBeVisible()
```
Comment (suzubara) [inline src/components/forms/ComboBox/ComboBox.test.tsx:812]: this is not the behavior I'm seeing.. in Chrome when I have selected Apricot and the Apple option is focused and I press the up arrow, the menu closes

File: src/components/forms/ComboBox/ComboBox.test.tsx [hunk 5]
```diff
@@ -0,0 +1,1111 @@
...
+      const { getByTestId } = render(
+        <ComboBox
+          id="favorite-fruit"
+          name="favorite-fruit"
+          options={fruitOptions}
+          onChange={jest.fn()}
+        />
+      )
+
+      fireEvent.click(getByTestId('combo-box-input'))
```
Comment (suzubara) [inline src/components/forms/ComboBox/ComboBox.test.tsx:889]: should this use the [dblclick event](https://developer.mozilla.org/en-US/docs/Web/API/Element/dblclick_event)?

File: src/components/forms/ComboBox/ComboBox.test.tsx [hunk 6]
```diff
@@ -0,0 +1,1111 @@
...
+
+      expect(getByTestId('combo-box-input')).toHaveAttribute(
+        'aria-expanded',
+        'true'
+      )
+      expect(getByTestId('combo-box-option-list')).toBeVisible()
+    })
+
+    it('hides options list when clicking away', () => {
```
Comment (suzubara) [inline src/components/forms/ComboBox/ComboBox.test.tsx:899]: should the options list be hidden if an option has focus and you click away? it hides when the input has focus and I click away but not if I've hovered over one of the options first

File: src/components/forms/ComboBox/ComboBox.test.tsx [hunk 7]
```diff
@@ -0,0 +1,1111 @@
...
+          options={fruitOptions}
+          onChange={jest.fn()}
+        />
+      )
+
+      const input = getByTestId('combo-box-input')
+      userEvent.type(input, 'yu')
+      userEvent.click(getByTestId('combo-box-option-yuzu'))
+
+      expect(getByTestId('combo-box-input')).toHaveValue('Yuzu')
```
Comment (suzubara) [inline src/components/forms/ComboBox/ComboBox.test.tsx:997]: It might be worth testing that the `onChange` handler was called with the value here

**Review by suzubara — COMMENTED** [review]
wow, this is an incredibly impressive PR and v1 of this component 👏  !! I think it is really solid and fine to follow up on the outstanding bugs as additional issues after merging. I left some comments, mostly noting differences between the tests and my UX manually testing in Chrome, and some questions.

The only thing I would consider a blocker is that I am seeing the selected value clear out whenever I blur, making it impossible to actually persist a value in the form (see gif). if we can fix that then I think this is good to go!

ComboBox bug

File: src/components/forms/ComboBox/ComboBox.test.tsx (hunk 2 above)
Comment (haworku) [reply src/components/forms/ComboBox/ComboBox.test.tsx:483]: Yes. This is documented in [a failing test](https://github.com/trussworks/react-uswds/blob/f6a08d3f59155e54c790f050c836a3826ff0c882/src/components/forms/ComboBox/ComboBox.test.tsx#L726-L746). Should it be a blocker? I'm thinking yes.

File: src/components/forms/ComboBox/ComboBox.test.tsx (hunk 3 above)
Comment (haworku) [reply src/components/forms/ComboBox/ComboBox.test.tsx:723]: Yes. This documented in [a failing test](https://github.com/trussworks/react-uswds/blob/f6a08d3f59155e54c790f050c836a3826ff0c882/src/components/forms/ComboBox/ComboBox.test.tsx#L706-L724).

File: src/components/forms/ComboBox/ComboBox.test.tsx (hunk 6 above)
Comment (haworku) [reply src/components/forms/ComboBox/ComboBox.test.tsx:899]: 🐛  That's a bug, will fix and write a test.

File: src/components/forms/ComboBox/ComboBox.test.tsx (hunk 4 above)
Comment (haworku) [reply src/components/forms/ComboBox/ComboBox.test.tsx:812]: 🐛 This is a bug and this test should be failing.  Changing `fireEvent.focus` to `userEvent.hover(getByTestId('combo-box-option-apple'))` makes it fail as it should.  Great catch!

**Comment by haworku** [comment]
> The only thing I would consider a blocker is that I am seeing the selected value clear out whenever I blur, making it impossible to actually persist a value in the form (see gif). if we can fix that then I think this is good to go!

Fixed in e0ebc24. Thanks!!

**Review by suzubara — APPROVED** [review]
Verified that the fixes look great in Storybook! They don't seem to be reflected in the example app but that is an existing issue. This is awesome!! 🚢  🚀  🔥  🎉 

(I took a look at the Happo diffs and they look fine, maybe will be resolved once this branch is updated from `main` or they could be a result of Happo updating their browsers)

**Review by suzubara — APPROVED** [review]

**Pull request merged**